### PR TITLE
Enhance Informer

### PR DESCRIPTION
**Pull request by rjbhewei**
1. informer list request support pagination like client-go(default limit 500).  

    eg: PageInformerExample.java

    fix issue : https://github.com/kubernetes-client/java/issues/628

2. informer local-cache filter resource fields which customer definition，to ease server-side JVM memory usage，avoid oom.

    eg: FieldFilterPageInformerExample.java

### CODE REVIEW

File: extended/pom.xml [hunk 1]
```diff
@@ -42,6 +42,12 @@
             <artifactId>bucket4j-core</artifactId>
             <version>${bucket4jVersion}</version>
         </dependency>
+        <dependency>
```
Comment (brendandburns) [inline extended/pom.xml]: These pom changes seem unrelated. Are they necessary?

File: extended/src/main/java/io/kubernetes/client/extended/informer/FilterField.java [hunk 2]
```diff
@@ -0,0 +1,48 @@
+package io.kubernetes.client.extended.informer;
+
+import org.apache.commons.lang3.StringUtils;
+
+public class FilterField {
+
+  // the name of the field
```
Comment (brendandburns) [inline extended/src/main/java/io/kubernetes/client/extended/informer/FilterField.java]: delete redundant comments. (here and below)

File: extended/src/main/java/io/kubernetes/client/extended/informer/FilterFunction.java [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+import java.util.List;
+import java.util.Objects;
+
+@FunctionalInterface
+public interface FilterFunction<T> {
+
+  T apply(T t);
+
+  default FilterFunction<T> compose(FilterFunction<T> before) {
```
Comment (brendandburns) [inline extended/src/main/java/io/kubernetes/client/extended/informer/FilterFunction.java]: I think these methods need JavaDoc

File: extended/src/main/java/io/kubernetes/client/extended/informer/PageSharedInformerFactory.java [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+
+        return pager.getListObjectCurrentPage();
+      }
+
+      @Override
+      public Watch<ApiType> watch(CallGeneratorParams callGeneratorParams) throws ApiException {
+
+        Call call = callGenerator.generate(PagerCallGeneratorParams.to(callGeneratorParams));
+
+        Type watchType = TypeToken.getParameterized(Watch.Response.class, apiTypeClass).getType();
```
Comment (brendandburns) [inline extended/src/main/java/io/kubernetes/client/extended/informer/PageSharedInformerFactory.java]: Perhaps move this to the constructor? It's never going to change.

File: extended/src/main/java/io/kubernetes/client/extended/informer/SerializeFilterJSON.java [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+import java.util.Map;
+import java.util.function.BiPredicate;
+import org.apache.commons.collections4.CollectionUtils;
+import org.apache.commons.collections4.ListUtils;
+import org.apache.commons.lang3.StringUtils;
+import org.joda.time.DateTime;
+import org.joda.time.LocalDate;
+
+/**
+ * enhance default JSON，which serialize list/watch objects，SerializeFilterJSON enhance common object
```
Comment (brendandburns) [inline extended/src/main/java/io/kubernetes/client/extended/informer/SerializeFilterJSON.java]: which serializes list/watch objects.

Comment (brendandburns) [reply extended/src/main/java/io/kubernetes/client/extended/informer/SerializeFilterJSON.java]: Clarify second sentence, it's hard to understand.

File: extended/src/main/java/io/kubernetes/client/extended/informer/SerializeFilterJSON.java [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+    if (CollectionUtils.isEmpty(filterFields)) {
+      return null;
+    }
+
+    return new ExclusionStrategy() {
+
+      @Override
+      public boolean shouldSkipField(FieldAttributes f) {
+
+        SerializedName serialized = f.getAnnotation(SerializedName.class);
```
Comment (brendandburns) [inline extended/src/main/java/io/kubernetes/client/extended/informer/SerializeFilterJSON.java]: This can be moved into the constructor right?

File: extended/src/main/java/io/kubernetes/client/extended/informer/SerializeFilterJSON.java [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+
+        FilterField filterField =
```
Comment (brendandburns) [inline extended/src/main/java/io/kubernetes/client/extended/informer/SerializeFilterJSON.java]: This too?

File: extended/src/main/java/io/kubernetes/client/extended/informer/SerializeFilterJSON.java [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+      @Override
+      public boolean shouldSkipClass(Class<?> clazz) {
+        return false;
+      }
+    };
+  }
+
+  private static final BiPredicate<FilterField, FilterField> MATCH_PREDICATE =
+      (customer, system) -> {
+        if (!customer.isNotEmpty()) { // double check
```
Comment (brendandburns) [inline extended/src/main/java/io/kubernetes/client/extended/informer/SerializeFilterJSON.java]: double negatives are hard to understand. why not `if (customer.isEmpty()) {`

File: extended/src/main/java/io/kubernetes/client/extended/informer/SerializeFilterJSON.java [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+          return false;
+        }
+
+        boolean isSameName =
+            StringUtils.isEmpty(customer.getName())
```
Comment (brendandburns) [inline extended/src/main/java/io/kubernetes/client/extended/informer/SerializeFilterJSON.java]: why is it the same if the customer name is empty? what if the system name is not?

File: extended/src/main/java/io/kubernetes/client/extended/informer/SerializeFilterJSON.java [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+
+      @Override
+      public boolean shouldSkipClass(Class<?> clazz) {
+        return false;
+      }
+    };
+  }
+
+  private static final BiPredicate<FilterField, FilterField> MATCH_PREDICATE =
+      (customer, system) -> {
```
Comment (brendandburns) [inline extended/src/main/java/io/kubernetes/client/extended/informer/SerializeFilterJSON.java]: I'm not sure why one is `customer` and one is `system`

File: util/src/main/java/io/kubernetes/client/util/CallGenerator.java [hunk 11]
```diff
@@ -13,13 +13,13 @@
  * params.resourceVersion, params.timeoutSeconds, params.watch, null, null); },
  */
 @FunctionalInterface
-public interface CallGenerator {
+public interface CallGenerator<T extends CallGeneratorParams> {
```
Comment (brendandburns) [inline util/src/main/java/io/kubernetes/client/util/CallGenerator.java]: Why is templating necessary here? It seems that polymorphism would take care of it.

**Comment by rjbhewei** [comment]
I wanted to enhance the new features by changing the original code as little as possible.Without affecting the original function。

original code use io.kubernetes.client.JSON serializes list/watch objects. but I need to ease server-side JVM memory usage，when serialize objects customer can definition filter some filter，so impl SerializeFilterJSON to do this

File: util/src/main/java/io/kubernetes/client/util/CallGenerator.java (hunk 11 above)
Comment (rjbhewei) [reply util/src/main/java/io/kubernetes/client/util/CallGenerator.java]: PageSharedInformerFactory use PagerCallGeneratorParams ,but ListerWatcher  list method hardcode  CallGeneratorParams

File: extended/src/main/java/io/kubernetes/client/extended/informer/FilterField.java (hunk 2 above)
Comment (rjbhewei) [reply extended/src/main/java/io/kubernetes/client/extended/informer/FilterField.java]: delete now

File: extended/src/main/java/io/kubernetes/client/extended/informer/SerializeFilterJSON.java (hunk 8 above)
Comment (rjbhewei) [reply extended/src/main/java/io/kubernetes/client/extended/informer/SerializeFilterJSON.java]: add isEmpty() method now

File: extended/src/main/java/io/kubernetes/client/extended/informer/SerializeFilterJSON.java (hunk 9 above)
Comment (rjbhewei) [reply extended/src/main/java/io/kubernetes/client/extended/informer/SerializeFilterJSON.java]: customer definition filterField  not use the "name" attribute，but use "declaringClass" or "serializedName"
